### PR TITLE
*: Update the UniqueID of partition expression columns in LIST partitioning

### PR DESCRIPTION
**Pull request by mjonss**
### What problem does this PR solve?

Issue Number: ref #42135

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by winoros — APPROVED** [review]
I prefer this one to be merged

**Comment by qw4990** [comment]
Hi Mattia, Please fix CI failures. @mjonss

**Comment by qw4990** [comment]
> Hi Mattia, Please fix CI failures. @mjonss

Better to let's merge this PR today. A customer is waiting for this 🤦

File: planner/core/casetest/testdata/integration_partition_suite_out.json [hunk 1]
```diff
@@ -615,7 +615,7 @@
         "SQL": "explain format = 'brief' insert into tcollist select * from tcollist where a in (2)",
         "DynamicPlan": [
           "Insert N/A root  N/A",
-          "└─TableReader 10.00 root partition:p0 data:Selection",
+          "└─TableReader 10.00 root partition:p0,p1,p2,p3 data:Selection",
```
Comment (qw4990) [inline planner/core/casetest/testdata/integration_partition_suite_out.json:618]: Weird... It seems like the partition pruning cannot work with `list column` partitioning tables.

File: planner/core/rule_partition_processor.go [hunk 2]
```diff
@@ -476,21 +476,21 @@ func (s *partitionProcessor) processHashOrKeyPartition(ds *DataSource, pi *model
...
-	colIDToUniqueID := make(map[int64]int64)
-	for _, cond := range conds {
-		condCols := expression.ExtractColumns(cond)
-		for _, c := range condCols {
-			colIDToUniqueID[c.ID] = c.UniqueID
+	s *partitionProcessor, conds []expression.Expression, pruneList *tables.ForListPruning, columns []*expression.Column) *listPartitionPruner {
+	for i := range pruneList.PruneExprCols {
+		for _, col := range columns {
+			if col.ID == pruneList.PruneExprCols[i].ID {
+				pruneList.PruneExprCols[i].UniqueID = col.UniqueID
```
Comment (qw4990) [inline planner/core/rule_partition_processor.go:491]: It seems like `pruneList` is from a global object, so if we update it here, will it cause some data race issues?

Comment (mjonss) [reply planner/core/rule_partition_processor.go:491]: Thank you for spotting it, I will try to clone it.

**Review by qw4990 — APPROVED** [review]
LGTM

**Pull request merged**